### PR TITLE
Cypress Docs

### PR DESCRIPTION
**Pull request by xoscar**
This PR adds the initial Cypress documentation

## Changes

- Adds Cypress docs

## Checklist

- [x] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

### CODE REVIEW

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+image: https://res.cloudinary.com/djwdcmwdz/image/upload/v1698686403/docs/Blog_Thumbnail_14_rsvkmo.jpg
+---
+
+[Tracetest](https://tracetest.io/) is a testing tool based on [OpenTelemetry](https://opentelemetry.io/) that permits you to test your distributed application. It allows you to use the trace data generated by your OpenTelemetry tools to check and assert if your application has the desired behavior defined by your test definitions.
+
+[Cypress](https://www.cypress.io/) is a JavaScript end-to-end testing framework. It is used for testing web applications by simulating user interactions within the browser. Cypress provides a fast, reliable, and easy-to-use testing environment for developers.
+
+## Why is this important?
+
+Cypress is currently one of the top e2e testing frameworks, it is a great tool in its own right that allows you to replicate most of the production challenges you might encounter by allowinig developers test the system from the user perspective. But, as with all of the tools that only test the UI portion application, you can only run validations against what the user would see while going through the flows.
```
Comment (kdhamric) [inline docs/docs/tools-and-integrations/cypress.mdx:21]: allowinig should be allowing

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+## Requirements
+
+**Tracetest Account**:
+
+- Sign up to [`app.tracetest.io`](https://app.tracetest.io) or follow the [get started](/getting-started/installation) docs.
+- Create an [environment](/concepts/environments).
+- Create an [environment token](/concepts/environment-tokens).
+- Have access to the environment's [agent API key](/configuration/agent).
+
+**Pokeshop Demo:** Clone the official Tracetest Pokeshop Demo App to your local machine.
```
Comment (kdhamric) [inline docs/docs/tools-and-integrations/cypress.mdx:54]: Can we provide a link to this repo?

File: docs/docs/tools-and-integrations/cypress.mdx (hunk 1 above)
Comment (jfermi) [inline docs/docs/tools-and-integrations/cypress.mdx:21]:
```suggestion
Cypress is currently one of the top e2e testing frameworks, it is a great tool in its own right that allows you to replicate most of the production challenges you might encounter by allowing developers to test the system from the user perspective. But, as with all of the tools that only test the UI portion application, you can only run validations against what the user would see while going through the flows.
```

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+  Scripts->>+Scripts: Visits website
+  Scripts->>+@tracetest/cypress: captures and injects traceparent meta tag
+  @tracetest/cypress-->-Scripts: Ok
+  Scripts->>+Scripts: Executes test
+  Scripts->>+@tracetest/cypress: Runs test
+  @tracetest/cypress-->-Scripts: Ok
+  Scripts->>@tracetest/cypress: Waits for results and shows the summary
+```
+
+## The `@tracetest/cypress` npm package
```
Comment (jfermi) [inline docs/docs/tools-and-integrations/cypress.mdx:41]:
```suggestion
## The `@tracetest/cypress` npm Package
```

**Review by kdhamric — COMMENTED** [review]
Marked a couple comments

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+## Project Structure
+
+The project is built with Docker Compose.
+
+### Pokeshop Demo App
+
+The [Pokeshop Demo App](/live-examples/pokeshop/overview) is a complete example of a distributed application using different backend and front-end services, implementation code is written in Typescript.
+
+The `docker-compose.yml` file in the root directory is for the Pokeshop Demo app and the OpenTelemetry setup.
+And the `docker-compose.e2e.yml` includes the and the [Tracetest Agent](/concepts/agent).
```
Comment (jfermi) [inline docs/docs/tools-and-integrations/cypress.mdx:67]: ...includes the ??? and the... Missing text after the first 'the'.

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+
+Finally, the Cypress E2E tests can be found in `cypress/e2e/1-getting-started`.
+
+## The Cypress setup
```
Comment (jfermi) [inline docs/docs/tools-and-integrations/cypress.mdx:71]:
```suggestion
## The Cypress Setup
```

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+      TRACETEST_API_TOKEN: process.env.TRACETEST_API_TOKEN, // used to authenticate with Tracetest
+    },
+    setupNodeEvents() {
+      // implement node event listeners here
+    },
+  },
+});
+```
+
+### The `home.cy` test script
```
Comment (jfermi) [inline docs/docs/tools-and-integrations/cypress.mdx:95]:
```suggestion
### The `home.cy` Test Script
```

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+
+The `home.cy` test script contains three different tests based on the Pokeshop Demo UI fetures, which are:
+
+1. Creates a Pokemon
+2. Imports a Pokemon (using an async process)
+3. Deletes a Pokemon
+
+### Tracetest library setup
```
Comment (jfermi) [inline docs/docs/tools-and-integrations/cypress.mdx:103]:
```suggestion
### Tracetest Library Setup
```

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+
+If you go to the `package.json` file you will find the inclusion of a Tracetest package for cypress `@tracetest/cypress`.
```
Comment (jfermi) [inline docs/docs/tools-and-integrations/cypress.mdx:105]:
```suggestion
If you go to the `package.json` file you will find the inclusion of a Tracetest package for Cypress `@tracetest/cypress`.
```

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+Lastly, during the `afterEach` hook, the script **runs** the test with an optional definition string.
+
+```typescript
+afterEach((done) => {
+  const definition = Cypress.env("definition") || "";
+  tracetest.runTest(definition).then(() => done());
+});
+```
+
+**OPTIONAL**: If you want to wait for the test to finish and break the cypress execution based on a failed Tracetest test, you can add the `after` hook and call the `summary` method.
```
Comment (jfermi) [inline docs/docs/tools-and-integrations/cypress.mdx:141]:
```suggestion
**OPTIONAL**: If you want to wait for the test to finish and break the Cypress execution based on a failed Tracetest test, you can add the `after` hook and call the `summary` method.
```

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 10]
```diff
@@ -0,0 +1,346 @@
...
+  });
+
+  it("deletes a pokemon (2)", () => {
+    cy.get('[data-cy="pokemon-list"]').should("be.visible");
+    cy.get('[data-cy="pokemon-card"]').first().click().get('[data-cy="delete-pokemon-button"]').first().click();
+  });
+});
+```
+
+### Setting the environment variables
```
Comment (jfermi) [inline docs/docs/tools-and-integrations/cypress.mdx:228]:
```suggestion
### Setting the Environment Variables
```

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 11]
```diff
@@ -0,0 +1,346 @@
...
+
+### Starting the Pokeshop Demo App
+
+To start the Pokeshop Demo App, run the following command from the root directory:
+
+```bash
+docker compose up -f docker-compose.yml -f docker-compose.e2e.yml
+```
+
+### Running the tests
```
Comment (jfermi) [inline docs/docs/tools-and-integrations/cypress.mdx:240]:
```suggestion
### Running the Tests
```

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 12]
```diff
@@ -0,0 +1,346 @@
...
+![Cypress UI](./img/cypress-scripts.png)
+
+You should see the three tests running and passing.
+And by looking at the browser console log you can find the statuses and results of the tests.
+
+![Cypress Results](./img/cypress-open-results.png)
+
+Then, you can follow any of the result links in the console log to the Tracetest App and see the results of the tests.
+
+The Tracetest library uses the same as the spec name for the trace-based tests, that way you can identify them more easily, it also fills some of the metadata directly from the cypress execution.
```
Comment (jfermi) [inline docs/docs/tools-and-integrations/cypress.mdx:337]:
```suggestion
The Tracetest library uses the spec name for the trace-based tests. That way you can identify them more easily and it also fills in some of the metadata directly from the Cypress execution. 
```

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 13]
```diff
@@ -0,0 +1,346 @@
...
+
+Then, you can follow any of the result links in the console log to the Tracetest App and see the results of the tests.
+
+The Tracetest library uses the same as the spec name for the trace-based tests, that way you can identify them more easily, it also fills some of the metadata directly from the cypress execution. 
+
+![Tracetest App](./img/cypress-tracetest-tests.png)
+
+Lastly, you can now create assertions based on the trace data that was captured from the browser to the backend services.
+
+Starting with the click events and the fetch request from the client side, to the HTTP requests and database queries from the backend services, including async processes like the one showcased during the import pokemon test.
```
Comment (jfermi) [inline docs/docs/tools-and-integrations/cypress.mdx:343]:
```suggestion
From the click events and the fetch request from the client side to the HTTP requests and database queries from the backend services, async processes like the one showcased during the import Pokemon test are included.
```

**Review by jfermi — COMMENTED** [review]
Some editing suggestions.

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 14]
```diff
@@ -0,0 +1,346 @@
+---
+id: cypress
+title: Running Tracetest with Cypress using the Pokeshop Demo
```
Comment (adnanrahic) [inline docs/docs/tools-and-integrations/cypress.mdx:3]:
```suggestion
title: Trace-Based End to End Testing with Cypress and Tracetest
```

File: docs/docs/tools-and-integrations/cypress.mdx [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+image: https://res.cloudinary.com/djwdcmwdz/image/upload/v1698686403/docs/Blog_Thumbnail_14_rsvkmo.jpg
+---
+
+[Tracetest](https://tracetest.io/) is a testing tool based on [OpenTelemetry](https://opentelemetry.io/) that permits you to test your distributed application. It allows you to use the trace data generated by your OpenTelemetry tools to check and assert if your application has the desired behavior defined by your test definitions.
+
+[Cypress](https://www.cypress.io/) is a JavaScript end-to-end testing framework. It is used for testing web applications by simulating user interactions within the browser. Cypress provides a fast, reliable, and easy-to-use testing environment for developers.
+
+## Why is this important?
+
+Cypress is currently one of the top e2e testing frameworks, it is a great tool in its own right that allows you to replicate most of the production challenges you might encounter by allowing developers to test the system from the user perspective. But, as with all of the tools that only test the UI portion application, you can only run validations against what the user would see while going through the flows.
```
Comment (adnanrahic) [inline docs/docs/tools-and-integrations/cypress.mdx:21]:
```suggestion
Cypress is currently one of the top end to end testing frameworks, it is a great tool in its own right that allows you to replicate most of the production challenges you might encounter by allowing developers to test the system from the user perspective. But, as with all of the tools that only test the UI portion application, you can only run validations against what the user would see while going through the flows.
```

**Review by adnanrahic — CHANGES_REQUESTED** [review]
Added small suggestions for SEO.

**Review by jorgeepc — APPROVED** [review]

**Pull request merged**